### PR TITLE
Update documentation of `ProgressBar` `out` parameter

### LINKED ISSUE #8524: ProgressBar `out` parameter issues
The `out` parameter in `diagnostics.ProgressBar` is not documented. It appears it should be a file-like object, which means bytes should be written to it by default. The current implementation raises `TypeError: a bytes-like object is required, not 'str'`.

Example:
```python
from tempfile import TemporaryFile

import dask.array as da
from dask.diagnostics import ProgressBar


with TemporaryFile() as temp_file:
    with ProgressBar(out=temp_file):
        da.random.random((10, 10)).compute()
```

Tested using:

- Dask version: 2021.12.0
- Python version: 3.9.7
- Operating System: Mac OS X 12.1
- Install method (conda, pip, source): conda

### PR DESCRIPTION
**Pull request by ppsbs**
Added documentation for the out parameter from diagnostics.ProgressBar

- [X] Closes #8524
- [X] Tests added / passed
- [X] Passes `pre-commit run --all-files`

cc @ian-r-rose

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by ppsbs** [comment]
ok to test

**Comment by jsignell** [comment]
ok to test

**Comment by jsignell** [comment]
Just FYI it has to be a code-owner who comments "ok to test" for GPU CI to start running, but I like your initiative @ppsbs :smile:

**Comment by ian-r-rose** [comment]
Thanks @ppsbs! I was about to comment that the linked issue came to the opposite conclusion -- that `out` couldn't be a file-like object. But I actually think that's not correct, and that your docstring here is okay! The issue in #8524 seems to be that `TemporaryFile` defaults to opening in binary mode, when we want to open the file in text mode.

All of which is to say, I'm happy with this as is

File: dask/diagnostics/progress.py [hunk 1]
```diff
@@ -37,6 +37,8 @@ class ProgressBar(Callback):
         Width of the bar
     dt : float, optional
         Update resolution in seconds, default is 0.1 seconds
+    out : file object, optional
+        File object to which the progress bar will be written. Default is sys.stdout.
```
Comment (ian-r-rose) [inline dask/diagnostics/progress.py:41]: `sys.stdout` could be enclosed in double backticks `` to use literal markup

Comment (ian-r-rose) [reply dask/diagnostics/progress.py:41]: Perhaps we could also add that the file needs to be opened in text mode (`"w"`) to address the point of confusion.

Comment (jsignell) [inline dask/diagnostics/progress.py:41]: I'm not sure if this explanation makes the behavior clearer, if you look at the original issue, passing a file object does not always work. The `out` parameter is really intended as a place where you can set `sys.stdout` or `sys.stderr`.

Comment (jsignell) [reply dask/diagnostics/progress.py:41]: Oops! I just saw that Ian has much more thorough feedback! Feel free to ignore this :point_up:

Comment (ppsbs) [reply dask/diagnostics/progress.py:41]: > Just FYI it has to be a code-owner who comments "ok to test" for GPU CI to start running, but I like your initiative @ppsbs 😄

that's so embarassing! Thanks for the tip @jsignell! :D

Comment (ppsbs) [reply dask/diagnostics/progress.py:41]: As @ian-r-rose , mentioned, out expects a file object in 'w' mode and TemporaryFile defaults to 'w+b' (https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryFile). But it's also true that, as you mentioned, from the previous modifications, it looks like only sys.stdout or sys.stderr were supposed to be used. 

I will think about it and update the fix. Thank you for your valuable tips @jsignell and @ian-r-rose !

**Comment by ppsbs** [comment]
I tried to take both comments into consideration! I hope now it's fine!
Thanks!

**Comment by jsignell** [comment]
This looks great @ppsbs! Thanks for doing this work (and for reminding me to merge it!).

**Comment by ppsbs** [comment]
> This looks great @ppsbs! Thanks for doing this work (and for reminding me to merge it!).

Thank you @jsignell ! It's my first merged PR in dask! :D

**Comment by jsignell** [comment]
Nice! Welcome :wave:

**Pull request merged**